feat: added event details component

prerana1821 opened the pull request:
### What is the change?

Created an event details component to show the user details of an event as described in the #112.

### Is it bug?
No 

### \*Dev Tested?

- [x] Yes
- [ ] No

### \*Tested on:

#### Platforms

- [x] Web

#### Browsers

- [x] Chrome
- [ ] Safari
- [x] Firefox

### Before / After Change Screenshots
![screencapture-localhost-3000-events-2023-04-15-16_31_05](https://user-images.githubusercontent.com/61601706/232211885-960490d3-c528-4900-9ac9-898153113819.png)

iRohitGaur commented:
@prerana1821 please address the merge conflicts

iRohitGaur commented on app/utils/interfaces.ts:
```diff
@@ -2,13 +2,15 @@ import { Event, stringOrDate } from 'react-big-calendar';
   };
 }
 export interface CalEvent extends Event {
   id?: number;
   description?: string;
   location?: string;
+  type?: 'public' | 'private';
   ownerId?: number;
   calendarId?: number;
   attendees?: Attendees[];
```
its `visibility` not `type`. It is already present in the latest `develop` branch. Please take a pull once

yesyash replied:
> its `visibility` not `type`. It is already present in the latest `develop` branch. Please take a pull once

Adding to this why have we kept `type` to be optional? 
Should it not be required always as an Event cannot be made without it? And we need it to display the event type inside the `event details` component.

prerana1821 replied:
Yes, fixed it. 

Note: `visibility` was also optional when I pull latest `develop ` branch.

iRohitGaur replied:
> Yes, fixed it.
> 
> Note: `visibility` was also optional when I pull latest `develop ` branch.

Thanks for the note. Everything is kept optional for an Event as it can be created with just the `title`, `from` and `to`. However, by default, an event will be private unless the user makes it public. So, we can go ahead and change `visibility` from optional to required. Can you please do that?

iRohitGaur commented on app/components/common/drawer/index.tsx:
```diff
@@ -0,0 +1,60 @@
+          </button>
+        </div>
+        <h1 className=" font-semibold text-3xl leading-10 capitalize">{event.title}</h1>
+        <span className="inline-flex items-center px-2.5 py-0.5 rounded-full text-sm font-medium bg-gray-100 text-gray-800 my-2">
+          <GlobeIcon style={{ color: '#737373' }} />
+          <p className="px-1">{event.type}</p>
+        </span>
```
we already have a component `EventVisibility`. You can use it here.

prerana1821 replied:
Thank you. Added it

iRohitGaur commented on app/components/Calendar.tsx:
```diff
@@ -5,27 +5,81 @@ import EventModal from '../components/common/eventModal';
+    type: 'public',
     start: moment().add(1, 'hour').toDate(),
     end: moment().add(2, 'hours').toDate(),
+    attendees: [
+      {
+        attendee: {
+          name: 'Ankush Dharkar',
+          email: 'attendee1@example.com',
+        },
+      },
+      {
+        attendee: {
+          name: 'Harshith Venkatesh',
+          email: 'attendee2@example.com',
+        },
+      },
+      {
+        attendee: {
+          name: 'Yash Raj',
+          email: 'attendee10@example.com',
+        },
+      },
+    ],
```
we just have the email of the attendee for now. We can display the emails instead of names.

- type needs to be replaced by visibility. Take the latest pull from `develop`, it already has the required data

yesyash replied:
> we just have the email of the attendee for now. We can display the emails instead of names.
> 
> * type needs to be replaced by visibility. Take the latest pull from `develop`, it already has the required data

Disagree here @iRohitGaur as displaying a user's email without their consent publicly is a bad idea. 
There are two ways I can think of to solve this:
1. Add a name field in the backend if not present already. 
2. If the backend change is a bigger task we can truncate the part in the email after `@` and show it as a username. 
Example: `attendee23@gmail.com` -> `attendee23`

iRohitGaur replied:
@yesyash What I have mentioned is how Google also maintains it. The `consent` part comes in our TOS and while user is creating/joining a public event.

yesyash replied:
After a discussion with @iRohitGaur, we came to the conclusion that the email Id will be hidden for all the events present in the `Social Events` tabs and visible in the `Events` tab where we show a user's private events. 

This is because, for a user's private event, there is a high chance the user will know other attendees of the event.

prerana1821 replied:
Thank you @yesyash and @iRohitGaur. I have fixed it.

yesyash commented on app/components/Calendar.tsx:
```diff
@@ -5,27 +5,81 @@ import EventModal from '../components/common/eventModal';
 // tilde operator not working so used relative imports
 
-const initialEventsList: CalEvent[] = [
+export const initialEventsList: CalEvent[] = [
```
@prerana1821 can we move this constant variable into another file like `components/Calendar.constants.ts` and change the variable name to `INITIAL_EVENTS_LIST` from `initialEventList` as this is a constant value and it's a common practice to name them in the capital?

iRohitGaur replied:
@yesyash there is already an issue created to work on it.

yesyash replied:
@iRohitGaur can you please link it here and resolve this?

yesyash commented on app/components/common/drawer/index.tsx:
```diff
@@ -0,0 +1,60 @@
+  toggleDrawer: () => void;
+}
+
+const Drawer: React.FC<DrawerProps> = ({ event, isDrawerVisible, toggleDrawer }) => (
```
Nitpick: 
Even if we do not have any logic inside this component, I think it's a good idea to do a regular return instead of this shorthand one to maintain consistency across the components. 

```typescript
const Drawer = (...) => {
    return (
        ...
    )
}
```

prerana1821 replied:
Yes, fixed it. Actually, I had added return prior but it was give me a eslint error.

yesyash commented on app/routes/events.tsx:
```diff
@@ -0,0 +1,40 @@
+
+  const toggleDrawer = (event?: CalEvent) => {
+    setSelectedEvent(event || null);
+    setIsDrawerVisible(!isDrawerVisible);
```
Nitpick: 
A better approach for this would be to pass an updater function `setIsDrawerVisible(prev => !prev)`  to maintain consistency and prevent any potential bugs if we add multiple updates in this toggleDrawer for some reason. 

Link to [react docs](https://react.dev/reference/react/useState#updating-state-based-on-the-previous-state)

prerana1821 replied:
Yes, thank you. Fixed it.

isVivek99 commented on app/components/common/drawer/index.tsx:
```diff
@@ -0,0 +1,60 @@
+        </div>
+        <h1 className=" font-semibold text-3xl leading-10 capitalize">{event.title}</h1>
+        <span className="inline-flex items-center px-2.5 py-0.5 rounded-full text-sm font-medium bg-gray-100 text-gray-800 my-2">
+          <GlobeIcon style={{ color: '#737373' }} />
```
we should move this color out, mostly in the tailwind.config

isVivek99 replied:
if we use EventVisibility component, please ignore this comment

isVivek99 reviewed (commented):
LGTM!

iRohitGaur reviewed (changes requested):
Few points to address:
- Mobile responsiveness
- Link the navbar `Events` to `/events` route

iRohitGaur commented on app/components/common/drawer/index.tsx:
```diff
@@ -0,0 +1,59 @@
+            </button>
+          </div>
+          <h1 className=" font-semibold text-3xl leading-10 capitalize">{event.title}</h1>
+          <EventVisibility visibility={event.visibility ?? ''} />
```
```suggestion
          <EventVisibility visibility={event.visibility} />
```
after you make `visibility` non-optional

prerana1821 commented:
> Few points to address:
> 
> * Mobile responsiveness
> * Link the navbar `Events` to `/events` route

Thank you, @iRohitGaur! Fixed it. 

![image](https://user-images.githubusercontent.com/61601706/235501227-d869416e-07cb-4424-9458-1eedd629c94e.png)

iRohitGaur commented:
> > Few points to address:
> > 
> > * Mobile responsiveness
> > * Link the navbar `Events` to `/events` route
> 
> Thank you, @iRohitGaur! Fixed it.
> 
> ![image](https://user-images.githubusercontent.com/61601706/235501227-d869416e-07cb-4424-9458-1eedd629c94e.png)

@prerana1821
- navbar `Events` still not pointing to `/events` route
- responsiveness still having issues with some sizes. Please check:
<img width="797" alt="Screenshot 2023-05-01 at 14 40 47" src="https://user-images.githubusercontent.com/10892977/235508735-f6ad7730-6638-4667-9148-a0dde98fb0f5.png">

prerana1821 commented:
> > > Few points to address:
> > > 
> > > * Mobile responsiveness
> > > * Link the navbar `Events` to `/events` route
> > 
> > 
> > Thank you, @iRohitGaur! Fixed it.
> > ![image](https://user-images.githubusercontent.com/61601706/235501227-d869416e-07cb-4424-9458-1eedd629c94e.png)
> 
> @prerana1821
> 
> * navbar `Events` still not pointing to `/events` route
> * responsiveness still having issues with some sizes. Please check:
> 
> <img alt="Screenshot 2023-05-01 at 14 40 47" width="797" src="https://user-images.githubusercontent.com/10892977/235508735-f6ad7730-6638-4667-9148-a0dde98fb0f5.png">

Fixed it! @iRohitGaur Please check

![image](https://user-images.githubusercontent.com/61601706/235584438-815c57c5-aa56-4990-b65b-22307f1a4a36.png)

iRohitGaur commented:
@prerana1821 I still see issues with other sizes.

1. The event details component is not scrollable:
![Screen Recording 2023-05-02 at 11 21 12](https://user-images.githubusercontent.com/10892977/235711581-df29e25b-e1f3-477a-946e-b15a776fd820.gif)

2. The rounded gray circle on the left of the email gets flattened when email is long for the view
<img width="206" alt="Screenshot 2023-05-02 at 11 24 10" src="https://user-images.githubusercontent.com/10892977/235711731-d08d0532-c72a-482c-8481-2f579c935239.png">

prerana1821 commented:
> @prerana1821 I still see issues with other sizes.
> 
> 1. The event details component is not scrollable:
>    ![Screen Recording 2023-05-02 at 11 21 12](https://user-images.githubusercontent.com/10892977/235711581-df29e25b-e1f3-477a-946e-b15a776fd820.gif)
>          
>            [
>              
>            
>                ![Screen Recording 2023-05-02 at 11 21 12](https://user-images.githubusercontent.com/10892977/235711581-df29e25b-e1f3-477a-946e-b15a776fd820.gif)
>              ](https://user-images.githubusercontent.com/10892977/235711581-df29e25b-e1f3-477a-946e-b15a776fd820.gif)
>            
>            
>              
>                
>                  
>                
>                
>                  
>                  
>                
>              
>              [
>                
>                  
>                
>              ](https://user-images.githubusercontent.com/10892977/235711581-df29e25b-e1f3-477a-946e-b15a776fd820.gif)
> 2. The rounded gray circle on the left of the email gets flattened when email is long for the view
> 
> <img alt="Screenshot 2023-05-02 at 11 24 10" width="206" src="https://user-images.githubusercontent.com/10892977/235711731-d08d0532-c72a-482c-8481-2f579c935239.png">

Hey @iRohitGaur,
Please share the settings in which you are facing this issue. 
Though I have made the component scrollable and tried fixing the email styling issue.
Please check & review.

iRohitGaur commented on app/components/common/drawer/index.tsx:
```diff
@@ -21,7 +21,7 @@ const Drawer: React.FC<DrawerProps> = ({ event, isDrawerVisible, toggleDrawer })
       <div
         onClick={toggleDrawer}
-        className={`pt-20 pb-12 px-12 sm:w-full md:w-1/2 lg:w-1/4 bg-white h-full absolute right-0 duration-300 ease-out transition-all flex flex-col justify-between ${
+        className={`pt-20 pb-12 px-12 sm:w-full md:w-1/2 lg:w-1/4 bg-white h-full absolute right-0 duration-300 ease-out transition-all flex flex-col justify-between overflow-scroll ${
```
can we also give a `min-w-[300px]` to this so that it doesn't squeeze too much?

prerana1821 replied:
Yes, added it. Thank you!

iRohitGaur commented:
> Hey @iRohitGaur, Please share the settings in which you are facing this issue. Though I have made the component scrollable and tried fixing the email styling issue. Please check & review.

@prerana1821 scrolling works now. Also the email styling issue is resolved.

prerana1821 commented:
> @prerana1821 right now all tests are failing. We merged a fix for it. Can you please take a pull and let me know so that we can merge this PR?

@iRohitGaur I have pulled the develop branch but two test cases are still failing.

iRohitGaur commented:
> @iRohitGaur I have pulled the develop branch but two test cases are still failing.

CC: @isVivek99

harshith-venkatesh commented:
https://stackoverflow.com/questions/70677061/syntaxerror-unexpected-token-export-when-using-fast-components-in-a-remix-app

harshith-venkatesh commented:
Hi @prerana1821 , Could you resolve this merge conflicts and raise the PR, you can add comment of those test cases which are failing
I will take those up

prerana1821 commented:
> Hi @prerana1821 , Could you resolve this merge conflicts and raise the PR, you can add comment of those test cases which are failing I will take those up

Hey @harshith-venkatesh, I have fixed the merge conflicts.

harshith-venkatesh commented:
Hi @prerana1821 , due to the on going discussion and testing helper that have been placed for Link in backlog, we can go ahead with the changes without the test cases
Ref: https://github.com/remix-run/remix/issues/4506
